api4/license.go

```diff
@@ -296,3 +299,123 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+	// Only admins can request a true up review.
+	if !c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionManageSystem) {
+		c.SetPermissionError(model.PermissionManageLicenseInformation)
+		return
+	}
+
+	license := c.App.Channels().License()
+	if license == nil {
+		http.Error(w, "A License is required to perform a true-up review", http.StatusBadRequest)
+		return
+	}
+
+	var subscription *model.Subscription
+	var err error
+	if c.App.Cloud() != nil {
+		// Subscription Data
+		userId := c.AppContext.Session().UserId
+		subscription, err = c.App.Cloud().GetSubscription(userId)
+		if err != nil || subscription == nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	seats := 0
+	if subscription != nil {
+		seats = subscription.Seats
+	}
+
+	// Customer Info & Usage Analytics
+	activeUserCount, err := c.App.Srv().Store().Status().GetTotalActiveUsersCount()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Webhook, calls, boards, and playbook counts
+	incomingWebhookCount, err := c.App.Srv().Store().Webhook().GetIncomingTotal()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	outgoingWebhookCount, err := c.App.Srv().Store().Webhook().GetOutgoingTotal()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Plugin Data
+	trueUpReviewPlugins := model.TrueUpReviewPlugins{
+		ActivePluginNames:   []string{},
+		InactivePluginNames: []string{},
+	}
+
+	if pluginResponse, err := c.App.GetPlugins(); err == nil {
+		for _, plugin := range pluginResponse.Active {
+			trueUpReviewPlugins.ActivePluginNames = append(trueUpReviewPlugins.ActivePluginNames, plugin.Name)
+			trueUpReviewPlugins.TotalActivePlugins += 1
+		}
+
+		for _, plugin := range pluginResponse.Inactive {
+			trueUpReviewPlugins.InactivePluginNames = append(trueUpReviewPlugins.InactivePluginNames, plugin.Name)
+			trueUpReviewPlugins.TotalInactivePlugins += 1
+		}
+	}
+
+	// Authentication Features
+	mfaUsed := c.App.Config().ServiceSettings.EnforceMultifactorAuthentication
+	ldapUsed := c.App.Config().LdapSettings.Enable
+	samlUsed := c.App.Config().SamlSettings.Enable
+	openIdUsed := c.App.Config().OpenIdSettings.Enable
+	guessAccessAllowed := c.App.Config().GuestAccountsSettings.Enable
+
+	authFeatures := map[string]*bool{
+		model.TrueUpReviewAuthFeaturesMfa:        mfaUsed,
+		model.TueUpReviewAuthFeaturesAdLdap:      ldapUsed,
+		model.TrueUpReviewauthFeaturesSaml:       samlUsed,
+		model.TrueUpReviewAuthFeatureOpenId:      openIdUsed,
+		model.TrueUpReviewAuthFeatureGuestAccess: guessAccessAllowed,
+	}
+
+	authFeatureList := []string{}
+	for feature, used := range authFeatures {
+		if used != nil && *used {
+			authFeatureList = append(authFeatureList, feature)
+		}
+	}
+
+	reviewProfile := model.TrueUpReviewProfile{
+		ServerId:               c.App.TelemetryId(),
```

Comment:
I like the well structured representation of this *internally* within the app. It looks like all existing telemetry is sent as a flat map of string to primitves. Its probably a good idea to do the same here to that its easier to consume for whatever human or code will be consuming this data.

We could still keep the structured internal representation but construct a flat map using a technique like this: https://github.com/mattermost/mattermost-server/blob/master/services/telemetry/telemetry.go#L869 .

Otherwise, if you feel strongly about sending out a structured map, I would make sure that the consumer will be easily able to read it.

Reply:
No strong feelings about a structured map. I think the only reason I ended up doing this was that I built all of this out _around_ the struct.

I definitely didn't think about the end game in terms of where the telemetry data is sent to. A flat map definitely sounds much safer, especially if that's what we're doing everywhere else. I'll make this change.